### ISSUE TITLE
Allow tab to be used to hide all UI, even when not in fullscreen mode

### ISSUE THREAD

**Issue by achadwick** [mypaint/mypaint#136]
Split out from #106. @AtsusaKaneytza requested:

-Pressing "Tab" used to toggle the display of subwindows in 1.0, but doesn't in 1.1. Would be helpful for maximizing drawing space.

**Comment by achadwick** [comment]
Deferring until after the 1.2 release. Rewiring this bit will just delay everything, and I'm kinda busy right now as it is.

Yeah, that's a decision to put off the decision.

**Comment by snan** [comment]
FWIW, I used to use this feature all the time and still miss it. I use dwm also.

**Comment by jplloyd** [comment]
What we have now is the ability to toggle auto-hiding of a static set of UI elements when in fullscreen.

What we don't have is the ability to:

* toggle hiding of extraneous elements when not in fullscreen
* configure which elements should be hidden by such a toggle 
(e.g. hide everything but the footer bar)

What we want to avoid is:

* **User confusion** - mentioned in #106, new users should not easily be able to instantly hide most of the UI without clearly knowing how they did it.
* **Full lock-out** - it should not be possible to hide the interface without the ability to bring them back. This could happen if the user is allowed to fully hide all menus in windowed mode, without keyboard/mouse bindings for neither the hide-toggle, fullscreen or the main popup menu. Yes, it is a contrived scenario, but possible.

To my mind, a sensible solution is to provide this functionality but either not giving it a shortcut by default or setting it up so that it doesn't hide the menu bar by default (was it hidden in 1.0.0? I don't remember).

The main thing I'm uncertain about is how this UI toggle should interact with auto-hiding in fullscreen.

What happens if:

1. Auto-hiding is on, UI toggle is on (hiding), user moves cursor to screen edge - should the relevant UI component show?
2. Toggle is then turned off - should UI reappear, overriding/disabling auto-hiding?

Basically, what should the relationship/precedence between an explicit hide-UI toggle and the auto-hide toggle be? What is sensible from a users point of view?

**Comment by snan** [comment]
DWM user (and heavy mypaint user) here.

I want tab to toggle it all and _just_ give me a canvas, even when not in fullscreen mode. The screen edges should be left alone i.e. _not_ bring up UI elements. Many version ago I could have a movie playing in one window, or chat or email or whatever, while painting in another window and having the paint window be just the canvas. And then toggle on and off if I need to pick a new color or brush or w/e.

I really really really miss this feature, the fullscreen (F11) mode is more than useless to me since it overrides all of the DWM panels. To me, bringing back tab-hiding, this is the biggest improvement mypaint could make, that and performance stuff.

I get that there is a confusion issue but it'd be so valuable to be able to get all of that stuff out of the way and just paint.

**Comment by TheMiziko** [comment]
@snan People who use my Paint probably were using photoshop, and in photoshop the ability to hide UI elements with Tab button is essential exactly for the reasons you mentioned. I do not see ANY reason to call it "User confusion" and I see a lot of posts asking for this feature.

**Comment by jplloyd** [comment]
First a couple of questions (requests for opinions):

How did/should enablling individual panels work with a hidden state. If hide-ui is on, and I enable e.g. the Layers Panel by pressing `L`, should it:

1. Do nothing - panel is marked as visible internally, but nothing is actually shown due to the hide-ui state.
2. Show the side bar where the panel is attached, but nothing else.
3. Show everything (disable hide-ui)

For option `2.` the question is if subsequent toggling should interpret the current state as hidden or not - i.e. should the next toggle after showing a panel:

1. Hide everything - even though the previous state was "hidden", everything wasn't actually hidden
2. Show everything - same behaviour as it would have normally.
--------
**TLDR**: proposed solution

* New configurable action "Show Canvas Only" (name stolen from Krita - maybe we can borrow their translations too :smile:  )
* Action is unbound by default (or bound to Tab, if enough people want that)
* State is not saved in user settings - always start with controls visible.
* If action is unbound and triggered via the menu, still display menu bar.
* When toggled on, it overrides fullscreen auto-hiding. This means that no UI controls will be shown when it is turned on, even when moving the cursor to the screen edge.
* If auto-hiding is on, and the new action is turned off, all UI controls are displayed for the same amount of time as they would be from edge-activation, then hidden again.
* Canvas overlays (same as for zoom messages) show the status to the user (w. information about relevant shortcuts).
* For display-panel behaviour, we go with options `2.` and sub-option `1.` above.

I think that seems ok, do point out if I've forgotten something.

--------

_Bit of rambling below, read at your own risk._

I've been mostly using a dwm-derivative window manager ([awesome wm](https://github.com/awesomeWM/awesome)) for the past 7 years, so I'm fully on board with the usefulness of this. What we shouldn't forget is that we are all experienced power-users compared to anyone who is just starting out.

> People who use my Paint probably were using photoshop,

That's possible, but a big assumption to make about new users.

> and in photoshop the ability to hide UI elements with Tab button is essential

Again, you may be right (about it being essential), though personally I didn't know about it (but I haven't really used Ps for 10 years, not sure it was even a feature back then).

>  I do not see ANY reason to call it "User confusion" and I see a lot of posts asking for this feature.

@TheMiziko Here's one scenario: 
1. New user has just learned that they can store and load brushes by using the number keys. 
2. New user, not looking at what they're doing, accidentally presses `Tab `instead of `1`. 
3. New user now cannot see any of the menus or panels. If the user does not realize that they pressed `Tab`, they will have to first realize that they can right click the canvas to bring up the main menu, and then search through that menu for the option to show the interface again. 
4. If they had previously bound right-click to e.g. panning, they will not even be able to open the menu.

Here's a worse scenario: 
1. User rebinds `Tab` to Color History or something, because they're familiar with that from some other software.
2. User rebinds right-click to e.g. Panning.
3. User accidentally selects "Hide UI" (or whatever it should be called) in a menu
4. User now has no interface, no menus, and no way to get them back without already knowing the relevant shortcuts (to e.g. the settings window - if that hasn't been rebound too).

That's why my initial instinct was to have this feature unbound (no shortcut assigned), so that experienced users who want this functionality, and understand how to use it, can rebind e.g. `Tab` to it as they wish. And also possibly that by default it should hide everything except the menu bar (but allow space-conscious users to easily configure it to hide everything).

And just to clarify the last sentence, by menu bar I don't mean the toolbar with the image buttons, just the bar with `File", "Edit", etc.

---------

Krita addresses this by using a notification when you enter their Canvas-Only mode that instructs you on how to leave it. However, if you happen to rebind `Tab` and enter the mode via the menu, it seems you're stuck in it until you try to quit the program (with unsaved changes, otherwise it just shuts down). I guess that's precedence for not worrying too much about the corner cases...

**Comment by snan** [comment]
> How did/should enablling individual panels work with a hidden state. If hide-ui is on, and I enable e.g. the Layers Panel by pressing L, should it:

Either is fine:  
1, 2.1, 2.2 or 3.  
In the world of 2.2 you just hit tab twice to make it all go away.

1 is what makes the most sense to me. 3 feels kinda pointless but enh, harmless enough. I can see real big pros to either of the two 2 approaches _if_ it works well (quickly popping up the correct panel and not doing it by mistake). It sorta undermines the whole idea of… _just_ drawing.

Also, of course I'm OK with this option not being bound by default and more something that we who really want it can enable. We do want to be careful about the new user experience. (That said, accidentally hitting F11 is just as "confusing" currently.)

**Comment by odysseywestra** [comment]
@achadwick Could you move this thread over to https://community.mypaint.org/c/feature-requests? I'm working on cleaning up the issue tracker by move all feature requests there so we can get community feedback. If you can also set your post as a wiki that way we can edit the scope of the feature request that would be awesome as well. Don't forget to link back to here as well so I can keep track of it incase we need to reopen the issue. Thanks.
